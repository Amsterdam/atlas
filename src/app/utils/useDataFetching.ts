import { useState } from 'react'
import { fetchWithToken } from '../../shared/services/api/api'

function useDataFetching<T = any>() {
  const [results, setResults] = useState<T | null>(null)
  const [errorMessage, setErrorMessage] = useState(false)
  const [loading, setLoading] = useState(false)

  async function fetchData(endpoint: string) {
    setLoading(true)
    try {
      const data = await fetchWithToken(endpoint)
      setResults(data)
    } catch (e) {
      setErrorMessage(e.message)
    }

    setLoading(false)
    return results
  }

  return {
    errorMessage,
    loading,
    results,
    fetchData,
  }
}

export default useDataFetching
