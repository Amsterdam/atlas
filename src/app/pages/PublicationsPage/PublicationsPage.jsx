import React from 'react'
import { connect } from 'react-redux'
import { Helmet } from 'react-helmet'
import download from 'downloadjs'
import {
  Column,
  Row,
  Spinner,
  Publication,
  CustomHTMLBlock,
  Summary,
  BlogHeader,
  BlogMetaList,
  DocumentCover,
  BlogContent,
  Container,
} from '@datapunt/asc-ui'
// import Footer from '../../components/Footer/Footer'
import SHARED_CONFIG from '../../../shared/services/shared-config/shared-config'
import { getLocationPayload } from '../../../store/redux-first-router/selectors'
import useDataFetching from '../../utils/useDataFetching'
import { toPublication } from '../../../store/redux-first-router/actions'
import getReduxLinkProps from '../../utils/getReduxLinkProps'
import normalizeFromCMS from '../../utils/normalizeFromCMS'
import { routing } from '../../routes'

/* istanbul ignore next */
const PublicationsPage = ({ id, endpoint }) => {
  const [publication, setPublication] = React.useState({})
  const { fetchData, results, loading } = useDataFetching()

  React.useEffect(() => {
    fetchData(endpoint)
  }, [])

  React.useEffect(() => {
    if (results) {
      try {
        const data = normalizeFromCMS(results, [
          'field_file_size',
          'field_file_type',
          'field_publication_source',
          'field_publication_intro',
          'field_slug',
        ])
  
        setPublication(data)
      } catch(e) {
        window.location.replace(routing.niet_gevonden.path)
      }
    }
  }, [results])

  const {
    title,
    localeDate,
    body,
    coverUrl,
    field_file_size: fileSize,
    field_file_type: fileType,
    field_publication_source: source,
    field_publication_intro: intro,
    field_slug: slug,
    included,
  } = publication

  const downloadUrl = included ? publication.included[3].attributes.uri.url : {}

  const action = toPublication(id, slug)
  const { href } = getReduxLinkProps(action)

  return (
    <div className="c-dashboard__page o-max-width">
      <Container>
        <Helmet>
          <link rel="canonical" href={href} />
        </Helmet>

        {loading && (
          <div className="loading-indicator">
            <Spinner size={36} color="secondary" />
          </div>
        )}
        <Row>
          <Column wrap="true" span={{ small: 1, medium: 4, big: 6, large: 12, xLarge: 12 }}>
            {!loading && body && (
              <Publication>
                <Row>
                  <Column wrap span={{ small: 1, medium: 4, big: 6, large: 12, xLarge: 12 }}>
                    <Column
                      span={{
                        small: 1,
                        medium: 4,
                        big: 6,
                        large: 12,
                        xLarge: 12,
                      }}
                    >
                      <BlogContent>
                        <BlogHeader title={title} />
                        <BlogMetaList
                          fields={[
                            { id: 1, label: source },
                            { id: 4, label: localeDate },
                            { id: 2, label: fileSize },
                            { id: 3, label: fileType.toUpperCase() },
                          ]}
                        />
                      </BlogContent>
                    </Column>
                    <Column span={{ small: 1, medium: 4, big: 3, large: 6, xLarge: 6 }}>
                      <DocumentCover
                        imageSrc={`${SHARED_CONFIG.CMS_ROOT}${coverUrl}`}
                        description={`Download PDF (${fileSize})`}
                        onClick={() => {
                          const link = `${SHARED_CONFIG.CMS_ROOT}${downloadUrl}`
                          download(link)
                        }}
                      />
                    </Column>
                    <Column span={{ small: 1, medium: 4, big: 3, large: 6, xLarge: 6 }}>
                      <BlogContent>
                        {intro && <Summary>{intro}</Summary>}
                        <CustomHTMLBlock body={body.value} />
                      </BlogContent>
                    </Column>
                  </Column>
                </Row>
              </Publication>
            )}
          </Column>
        </Row>
      </Container>
      {/* {!loading && <Footer />} */}
    </div>
  )
}

const mapStateToProps = state => {
  const { id } = getLocationPayload(state)
  return {
    id,
    endpoint: `${
      SHARED_CONFIG.CMS_ROOT
    }jsonapi/node/publication?filter[drupal_internal__nid]=${id}&include=field_cover_image.field_media_image,field_file.field_media_file`,
  }
}

export default connect(
  mapStateToProps,
  null,
)(PublicationsPage)
