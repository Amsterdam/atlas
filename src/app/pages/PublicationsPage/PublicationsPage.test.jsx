import React from 'react'
import { mount, shallow, render } from 'enzyme'
import configureMockStore from 'redux-mock-store'
import { ThemeProvider } from '@datapunt/asc-ui'
import SHARED_CONFIG from '../../../shared/services/shared-config/shared-config'
import PublicationsPage from './PublicationsPage'
import useDataFetching from '../../utils/useDataFetching'

jest.mock('../../utils/useDataFetching')

describe('PublicationsPage', () => {
  const id = 3

  const mockData = {
    fetchData: jest.fn(),
    results: {
      data: [
        {
          attributes: {
            title: 'This is a title',
            created: '2015-05-05',
            body: {
              processed: 'body text',
            },
            field_file_size: 'file size',
            field_file_type: 'pdf',
            field_publication_source: 'source',
            field_publication_intro: 'intro',
          },
        },
      ],
      included: [
        { attributes: { uri: { url: 'https://cover-link' } } },
        { attributes: { uri: { url: 'https://document-link' } } },
      ],
    },
  }

  let store
  beforeEach(() => {
    store = configureMockStore()({ location: { payload: { id } } })
  })

  it('should render the spinner when the request is loading', () => {
    useDataFetching.mockImplementation(() => ({
      loading: true,
    }))

    const component = shallow(<PublicationsPage />, {
      context: { store },
    }).dive()

    const spinner = component.find('Spinner').at(0)
    expect(spinner.exists()).toBeTruthy()
  })

  it('should call the fetchData function when the component mounts', () => {
    const fetchDataMock = jest.fn()
    useDataFetching.mockImplementation(() => ({
      fetchData: fetchDataMock,
      loading: true,
    }))

    const component = mount(
      <ThemeProvider>
        <PublicationsPage store={store} />
      </ThemeProvider>,
      { context: { store } },
    )

    const endpoint = `${SHARED_CONFIG.CMS_ROOT}jsonapi/node/publication?filter[drupal_internal__nid]=${id}&include=field_cover_image,field_publication_file`

    expect(component.find('PublicationsPage').props().endpoint).toBe(endpoint)

    expect(fetchDataMock).toHaveBeenCalledWith(endpoint)
  })

  it('should render the publication when there are results', () => {
    useDataFetching.mockImplementation(() => mockData)
    // console.log(store)
    const component = render(
      <ThemeProvider>
        <PublicationsPage store={store} />
      </ThemeProvider>,
      { context: { store } },
    )
    expect(component).toMatchSnapshot()
  })
})
