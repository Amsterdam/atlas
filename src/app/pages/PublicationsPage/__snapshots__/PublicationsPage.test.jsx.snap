// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PublicationsPage should render the publication when there are results 1`] = `
<div
  class="c-dashboard__page o-max-width"
>
  <div
    class="sc-jhAzac hhZpgF"
  >
    <div
      class="sc-fBuWsC cpXXyN"
    >
      <div
        class="sc-fMiknA duARQi"
      >
        <div
          class="sc-hzDkRC gFrWgH"
        />
      </div>
    </div>
  </div>
</div>
`;
